hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastOperandTypes.java

```diff
@@ -0,0 +1,175 @@
+import static org.apache.calcite.sql.type.SqlTypeName.NULL;
+
+/**
+ * A collection of operand type checkers. Basically, a mirror of {@link
+ * OperandTypes} provided by Calcite with various enhancements.
+ */
+public final class HazelcastOperandTypes {
+
+    /**
+     * The same as Calcite's {@link OperandTypes#COMPARABLE_ORDERED_COMPARABLE_ORDERED},
+     * but selects the least restrictive type as a common type. We do character
+     * coercion provided by {@link SqlOperandTypeChecker.Consistency#COMPARE} and
+     * used by Calcite on our own.
+     */
+    public static final SqlOperandTypeChecker COMPARABLE_ORDERED_COMPARABLE_ORDERED =
+            new ComparableOperandTypeChecker(2, RelDataTypeComparability.ALL,
+                    SqlOperandTypeChecker.Consistency.LEAST_RESTRICTIVE);
+
+    private HazelcastOperandTypes() {
+    }
+
+    /**
+     * @return the base operand type checker wrapped into a new type checker
+     * disallowing ANY type.
+     */
+    public static SqlSingleOperandTypeChecker notAny(SqlOperandTypeChecker base) {
+        return new NotAny(base);
+    }
+
+    /**
+     * @return the base operand type checker wrapped into a new type checker
+     * disallowing all of the operands to be of NULL type simultaneously.
+     */
+    public static SqlOperandTypeChecker notAllNull(SqlOperandTypeChecker base) {
+        return new NotAllNull(base);
+    }
+
+    private static final class NotAny implements SqlSingleOperandTypeChecker {
+
+        private final SqlOperandTypeChecker base;
+
+        NotAny(SqlOperandTypeChecker base) {
+            this.base = base;
+        }
+
+        @Override
+        public boolean checkOperandTypes(SqlCallBinding binding, boolean throwOnFailure) {
+            if (!base.checkOperandTypes(binding, throwOnFailure)) {
+                return false;
+            }
+
+            for (int i = 0; i < binding.getOperandCount(); ++i) {
+                if (!checkSingleOperandType(binding, binding.operand(i), i, throwOnFailure)) {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        @Override
+        public SqlOperandCountRange getOperandCountRange() {
+            return base.getOperandCountRange();
+        }
+
+        @Override
+        public String getAllowedSignatures(SqlOperator op, String opName) {
+            return base.getAllowedSignatures(op, opName);
+        }
+
+        @Override
+        public Consistency getConsistency() {
+            return base.getConsistency();
+        }
+
+        @Override
+        public boolean isOptional(int i) {
+            return base.isOptional(i);
+        }
+
+        @Override
+        public boolean checkSingleOperandType(SqlCallBinding binding, SqlNode operand, int index, boolean throwOnFailure) {
+            if (typeName(binding.getOperandType(index)) == ANY) {
+                if (throwOnFailure) {
+                    throw binding.newValidationSignatureError();
```

Comment:
The same problem as with "not all null" - the error message contains invalid "supported forms".

Reply:
created an issue https://github.com/hazelcast/hazelcast/issues/17301 for 4.2